File: 백준/Gold/2042. 구간 합 구하기/구간 합 구하기.py
import sys


def init(start, end, index):
    if start == end:
        segment_tree[index] = nums[start - 1]
        return segment_tree[index]
    mid = (start + end) // 2
    segment_tree[index] = init(start, mid, 2 * index) + init(mid + 1, end, 2 * index + 1)
    return segment_tree[index]


def find(start, end, index, left, right):
    if start > right or end < left:
        return 0
    if start >= left and end <= right:
        return segment_tree[index]
    mid = (start + end) // 2
    return find(start, mid, 2 * index, left, right) + find(mid + 1, end, 2 * index + 1, left, right)


def update(start, end, index, update_index, data):
    if start > update_index or end < update_index:
        return
    segment_tree[index] += data
    if start == end:
        return
    mid = (start + end) // 2
    update(start, mid, 2 * index, update_index, data)
    update(mid + 1, end, 2 * index + 1, update_index, data)


n, m, k = map(int, sys.stdin.readline().split())
nums = [int(sys.stdin.readline()) for _ in range(n)]
segment_tree = [0] * (n * 4)
init(1, n, 1)
for _ in range(m + k):
    a, b, c = map(int, sys.stdin.readline().split())
    if a == 1:
        num = c - nums[b - 1]
        nums[b - 1] = c
        update(1, n, 1, b, num)
    if a == 2:
        print(find(1, n, 1, b, c))

File: 백준/Gold/2042. 구간 합 구하기/구간 합 구하기.java
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.StringTokenizer;

public class Main {

	static BufferedReader br;
	static StringBuilder sb;
	static StringTokenizer st;

	static int n, m, k;
	static long[] nums;
	static long[] segmentTree;

	public static void main(String[] args) throws Exception {
		br = new BufferedReader(new InputStreamReader(System.in));
		sb = new StringBuilder();
		st = new StringTokenizer(br.readLine());
		n = Integer.parseInt(st.nextToken());
		m = Integer.parseInt(st.nextToken());
		k = Integer.parseInt(st.nextToken());
		nums = new long[n];
		for (int i = 0; i < n; i++) {
			nums[i] = Long.parseLong(br.readLine());
		}
		segmentTree = new long[n * 4];
		init(1, n, 1);
		for (int i = 0; i < m + k; i++) {
			st = new StringTokenizer(br.readLine());
			final int a = Integer.parseInt(st.nextToken());
			final int b = Integer.parseInt(st.nextToken());
			final long c = Long.parseLong(st.nextToken());
			if (a == 1) {
				final long num = c - nums[b - 1];
				nums[b - 1] = c;
				update(1, n, 1, b, num);
			}
			if (a == 2) {
				sb.append(find(1, n, 1, b, (int) c))
						.append("\n");
			}
		}
		System.out.println(sb);
	}

	private static long init(final int start, final int end, final int index) {
		if (start == end) {
			segmentTree[index] = nums[start - 1];
			return segmentTree[index];
		}
		final int mid = (start + end) / 2;
		segmentTree[index] = init(start, mid, 2 * index) + init(mid + 1, end, 2 * index + 1);
		return segmentTree[index];
	}

	private static long find(final int start, final int end, final int index, final int left, final int right) {
		if (start > right || end < left) {
			return 0;
		}
		if (start >= left && end <= right) {
			return segmentTree[index];
		}
		final int mid = (start + end) / 2;
		return find(start, mid, 2 * index, left, right) + find(mid + 1, end, 2 * index + 1, left, right);
	}

	private static void update(final int start, final int end, final int index, final int updateIndex,
			final long data) {
		if (start > updateIndex || end < updateIndex) {
			return;
		}
		segmentTree[index] += data;
		if (start == end) {
			return;
		}
		final int mid = (start + end) / 2;
		update(start, mid, 2 * index, updateIndex, data);
		update(mid + 1, end, 2 * index + 1, updateIndex, data);
	}
}
